### PR TITLE
Add support for SockJS

### LINKED ISSUE #6878: Native build failing when using vertx SockJSHandler
**Describe the bug**
Getting that error when packaging in native:
```
Error: Classes that should be initialized at run time got initialized during image building:
 io.netty.buffer.UnpooledHeapByteBuf the class was requested to be initialized at build time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf).  To see why io.netty.buffer.UnpooledHeapByteBuf got initialized use -H:+TraceClassInitializationio.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf the class was requested to be initialized at build time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf).  To see why io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf got initialized use -H:+TraceClassInitializationio.netty.buffer.UnpooledUnsafeHeapByteBuf the class was requested to be initialized at build time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf).  To see why io.netty.buffer.UnpooledUnsafeHeapByteBuf got initialized use -H:+TraceClassInitializationio.netty.buffer.AbstractReferenceCountedByteBuf the class was requested to be initialized at build time (from the command line).  To see why io.netty.buffer.AbstractReferenceCountedByteBuf got initialized use -H:+TraceClassInitialization
com.oracle.svm.core.util.UserError$UserException: Classes that should be initialized at run time got initialized during image building:
 io.netty.buffer.UnpooledHeapByteBuf the class was requested to be initialized at build time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf).  To see why io.netty.buffer.UnpooledHeapByteBuf got initialized use -H:+TraceClassInitializationio.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf the class was requested to be initialized at build time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf).  To see why io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf got initialized use -H:+TraceClassInitializationio.netty.buffer.UnpooledUnsafeHeapByteBuf the class was requested to be initialized at build time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf).  To see why io.netty.buffer.UnpooledUnsafeHeapByteBuf got initialized use -H:+TraceClassInitializationio.netty.buffer.AbstractReferenceCountedByteBuf the class was requested to be initialized at build time (from the command line).  To see why io.netty.buffer.AbstractReferenceCountedByteBuf got initialized use -H:+TraceClassInitialization
        at com.oracle.svm.core.util.UserError.abort(UserError.java:65)
        at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.checkDelayedInitialization(ConfigurableClassInitialization.java:494)
        at com.oracle.svm.hosted.classinitialization.ClassInitializationFeature.duringAnalysis(ClassInitializationFeature.java:188)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$8(NativeImageGenerator.java:711)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:63)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:711)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:526)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:444)
        at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

Here is the code that makes it fail:
```
 public void init(@Observes Router router) {
        BridgeOptions opts = new BridgeOptions()
            .addOutboundPermitted(new PermittedOptions().setAddress("displayGameObject"))
            .addOutboundPermitted(new PermittedOptions().setAddress("removeGameObject"))
            .addInboundPermitted(new PermittedOptions().setAddress("init-session"))
            .addInboundPermitted(new PermittedOptions().setAddress("play"))
            .addInboundPermitted(new PermittedOptions().setAddress("pause"))
            .addInboundPermitted(new PermittedOptions().setAddress("reset"));

        // Create the event bus bridge and add it to the router.
        SockJSHandler sockJSHandler = SockJSHandler.create(vertx);
        sockJSHandler.bridge(opts);
        router.route("/eventbus/*").handler(sockJSHandler);

        eventBus.consumer("play", msg -> publishGameEvent("play"));
        eventBus.consumer("pause", msg -> publishGameEvent("pause"));
        eventBus.consumer("reset", msg -> publishGameEvent("reset"));

    }
```

And here are the lines that when removed make it pass:
```
   // Create the event bus bridge and add it to the router.
        SockJSHandler sockJSHandler = SockJSHandler.create(vertx);
        sockJSHandler.bridge(opts);
        router.route("/eventbus/*").handler(sockJSHandler);
```

**Environment (please complete the following information):**
 - Output of `java -version`: 
```
openjdk version "1.8.0_232"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_232-b09)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.232-b09, mixed mode)
```
 - GraalVM version (if different from Java): 19.2.1
 - Quarkus version or git rev: 1.2.0.Final

### PR DESCRIPTION
**Pull request by cescoffier**
Some classes must not be initialized during the build, only at runtime.

This allows using SockJS and the SockJS bridge (see https://github.com/quarkusio/quarkus-quickstarts/pull/445 as an example)

### CODE REVIEW

**Review by stuartwdouglas — APPROVED** [review]
Content looks good to me, but I don't understand the import order changes?

**Review by machi1990 — DISMISSED** [review]

**Comment by machi1990** [comment]
>  but I don't understand the import order changes?

Probably the cause of CI failure

File: extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java [hunk 1]
```diff
@@ -158,4 +161,12 @@ ServiceStartBuildItem finalizeRouter(
 
         return new ServiceStartBuildItem("vertx-http");
     }
+
+    @BuildStep
+    NativeImageConfigBuildItem build(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
```
[the remark this answers is not in the document]
Comment (cescoffier) [reply extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java]: Fixed, my bad... copy the signature form another extension.

**Comment by cescoffier** [comment]
I've fixed the import sorting. I still don't get why it generates the import in the wrong order even when using the `-Pformat` in maven.

**Comment by ia3andy** [comment]
I am testing it with the failing code now ;-) thanks @cescoffier

**Comment by ia3andy** [comment]
@cescoffier maybe I did something wrong, but I get this when trying a native build with the PR:
```
 Failed to build quarkus application: java.lang.ClassNotFoundException: io.quarkus.deployment.steps.DeploymentClassLoaderBuildStep 
```

**Comment by ia3andy** [comment]
@cescoffier ok the error was a local maven cache problem.

Now I have a real error during native build:
```
java.lang.ClassNotFoundException: io.vertx.ext.web.client.impl.MultipartFormUpload
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at io.quarkus.runner.AutoFeature.beforeAnalysis(AutoFeature.zig:363)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$7(NativeImageGenerator.java:669)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:63)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:669)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:530)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:445)
        at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
java.lang.NullPointerException
        at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.initializeAtRunTime(ConfigurableClassInitialization.java:249)
```

**Review by ia3andy — CHANGES_REQUESTED** [review]
@cescoffier I have a new error now during native build:
```
java.lang.ClassNotFoundException: io.vertx.ext.web.client.impl.MultipartFormUpload
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at io.quarkus.runner.AutoFeature.beforeAnalysis(AutoFeature.zig:363)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$7(NativeImageGenerator.java:669)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:63)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:669)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:530)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:445)
        at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
java.lang.NullPointerException
        at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.initializeAtRunTime(ConfigurableClassInitialization.java:249)
```

**Comment by cescoffier** [comment]
Did you add quarkus-vertx-web as a dependency? Wondering if that's the
issue.

Le ven. 31 janv. 2020 à 10:30, Andy Damevin <notifications@github.com> a
écrit :

**Comment by ia3andy** [comment]
> Did you add quarkus-vertx-web as a dependency? Wondering if that's the issue.

@cescoffier yes I did

**Review by ia3andy — APPROVED** [review]

**Comment by cescoffier** [comment]
@ia3andy found a mistake in the code. On of the declared class comes from another Vert.x dependency. It needed to be removed.

**Review by machi1990 — APPROVED** [review]

File: extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java [hunk 2]
```diff
@@ -158,4 +161,11 @@ ServiceStartBuildItem finalizeRouter(
 
         return new ServiceStartBuildItem("vertx-http");
     }
+
+    @BuildStep
+    NativeImageConfigBuildItem build() {
+        return NativeImageConfigBuildItem.builder()
+                .addRuntimeInitializedClass("io.vertx.ext.web.handler.sockjs.impl.XhrTransport")
```
Comment (gsmet) [inline extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java]: Why not use `RuntimeInitializedClassBuildItem` instead?

**Review by gsmet — CHANGES_REQUESTED** [review]
Suggested a simplification.

File: extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java (hunk 2 above)
Comment (cescoffier) [reply extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java]: Fixed.

**Review by gsmet — APPROVED** [review]

**Pull request merged**